packages/neuron-wallet/tests/utils/export-history/index.test.ts

```diff
@@ -1,29 +1,147 @@
+import fs from 'fs'
+import path from 'path'
+import { initConnection, closeConnection, saveTransactions } from '../../setupAndTeardown/index'
+import transactions from '../../setupAndTeardown/transactions.fixture'
 import i18n from '../../../src/locales/i18n'
-import TransactionPersistor from '../../../src/services/tx/transaction-persistor'
-import { initConnection } from '../../../src/database/chain/ormconfig'
-import Transaction from '../../../src/models/chain/transaction'
-import fixtures from './fixtures.json'
-
-describe.skip('Test exporting history', () => {
-  // TODO: reuse test cases of TransactionsService#getAllByAddresses
-  beforeAll(async () => {
+import AddressService from '../../../src/services/addresses'
+import exportHistory from '../../../src/utils/export-history'
+
+describe('Test exporting history', () => {
+  const WALLET_ID = 'wallet id'
+  const MAINNET_CHAIN_TYPE = 'ckb'
+  const TESTNET_CHAIN_TYPE = 'ckb_testnet'
+  const FILE_PATH = path.join(os.tmpdir(), 'transaction.csv')
+  const stubProvider = {
+    walletID: WALLET_ID,
+    chainType: MAINNET_CHAIN_TYPE,
+    filePath: FILE_PATH,
+  }
+
+  beforeAll(() => {
     i18n.changeLanguage('en')
-    await initConnection(":memory:")
-    for (const fixture of Object.values(fixtures.transactions)) {
-      const tx = Transaction.fromSDK(
-        fixture.transaction as any,
-        fixture.transaction as any,
-      )
-      tx.outputs.forEach((o, i) => o.daoData = (fixture.transaction.outputs[i] as any).daoData)
-      await TransactionPersistor.saveFetchTx(tx)
+  })
+
+  beforeEach(async () => {
+    await initConnection()
+    return saveTransactions(transactions)
+  })
+
+  afterEach(() => {
+    try {
+      fs.unlinkSync(stubProvider.filePath)
+    } catch {
+      // ignore
     }
+    stubProvider.walletID = WALLET_ID
+    stubProvider.chainType = MAINNET_CHAIN_TYPE
+    stubProvider.filePath = FILE_PATH
+    return closeConnection()
   })
 
-  afterAll(() => {
-    return getConnection().close()
+  describe('When wallet id is missing', () => {
+    beforeEach(() => {
+      stubProvider.walletID = ''
+    })
+
+    it("Should throw an error", async () => {
+      expect.assertions(1)
+      try {
+        await exportHistory(stubProvider)
+      } catch (err) {
+        expect(err).toEqual(new Error('Wallet ID is required'))
+      }
+    })
+  })
+
+  describe('When filePath is missing', () => {
+    beforeEach(() => {
+      stubProvider.filePath = ''
+    })
+
+    it('Should throw an error', async () => {
+      expect.assertions(1)
+      try {
+        await exportHistory(stubProvider)
+      } catch (err) {
+        expect(err).toEqual(new Error('File Path is required'))
+      }
+    })
   })
 
-  it.skip('Test export to csv', async () => {
+  describe('When file exists', () => {
+    const originalExistsSync = fs.existsSync
+    const originalUnlinkSync = fs.unlinkSync
+
+    afterAll(() => {
+      fs.existsSync = originalExistsSync
```

Comment:
The cleaner way to mock this is via the `jest.doMock`, which can be used to mock the `fs` dependency when loading `exportHistory`. So that we don't need to restore the functions by ourselves. 

https://jestjs.io/docs/en/jest-object#jestdomockmodulename-factory-options

Reply:
I changed the code into
```js
describe('When file exists', () => {
    let mockFs: any
    beforeEach(() => {
      mockFs = jest.doMock('fs', () => ({
        __esModule: true,
        default: {
          existsSync: jest.fn().mockReturnValue(true),
          unlinkSync: jest.fn().mockReturnValue(undefined),
        },
        existsSync: jest.fn().mockReturnValue(true),
        unlinkSync: jest.fn().mockReturnValue(undefined),
      }))
    })

    it('Should remove file first', async () => {
      expect.assertions(2)
      await exportHistory(stubProvider)
      expect(mockFs.existsSync).toHaveBeenCalledWith(stubProvider.filePath)
      expect(mockFs.unlinkSync).toHaveBeenCalledWith(stubProvider.filePath)
    })
})
```
But the tests failed and `mockFs.existsSync` is `undefined`

Is there anything missing?

Reply:
And I've tried
```js
describe('When file exists', () => {
    jest.doMock('fs', () => {
      return {
        existsSync: jest.fn().mockReturnValue(true),
        unlinkSync: jest.fn().mockReturnValue(undefined)
      }
    })

    it('Should remove file first', async () => {
      expect.assertions(2)
      await exportHistory(stubProvider)
      expect(fs.existsSync).toHaveBeenCalledWith(stubProvider.filePath)
      expect(fs.unlinkSync).toHaveBeenCalledWith(stubProvider.filePath)
    })
```
Doesn't work either, and the following tests are broken.

Reply:
I think it is because the `exportHistory` has to be loaded after the `jest.doMock`.
Here is an example: https://github.com/nervosnetwork/neuron/blob/72fdf037fb5898c81222e24efde9633dfe958c9c/packages/neuron-wallet/tests/block-sync-render/queue.intg.test.ts#L180

Reply:
Do I need to import it again each time I define the mock? It seems quite weird... I'll have a try.

Reply:
You should just need to import once. Normally you defer the import until all the necessary mocks via `jest.doMock` has been set up.

Reply:
```js
  describe('When file exists', () => {
    beforeAll(() => {
      jest.doMock('fs', () => ({
        existsSync: jest.fn().mockReturnValue(true),
        unlinkSync: jest.fn().mockReturnValue(undefined)
      }))
    })

    it('Should remove file first', async () => {
      
      expect.assertions(2)
      const fn = require('../../../src/utils/export-history').default
      const mockFs = require('fs')
      await fn(stubProvider)
      expect(mockFs.existsSync).toHaveBeenCalledWith(stubProvider.filePath)
      expect(mockFs.unlinkSync).toHaveBeenCalledWith(stubProvider.filePath)
    })
  })
```
Still not work, could you do me a favor to fix the test with `doMock`?

And one more question, if I want to define a series of returned value of the mocked function for each case, should I define it at the top of test block with `mockReturnValueOne(A).mockReturnValue(B)...` and then import the module to test instead of defining the mock in each test case independently?

Reply:
Sure, no worries. I will take a look at this later today and try to come up with some examples of doing the mocks for this case.

Reply:
Actually no needs to use `jest.doMock` here for fs module. `jest.spyOn` is sufficient. Please see PR https://github.com/nervosnetwork/neuron/pull/1741

Reply:
`spyOn` does the same work here and can be restored with `jest.restoreAllMocks` manually, but what's the benefit of using `spyOn` instead of `fn`?

IMO, `spyOn` is used to assert the behavior of calls, and expected to return the real result instead of a mock result. And the docs also suggests that use `jest.spyOn(object, methodName).mockImplementation(() => customImplementation)` or `object[methodName] = jest.fn(() => customImplementation)` to skip the real invocation.

Reply:
> spyOn does the same work here and can be restored with jest.restoreAllMocks manually, but what's the benefit of using spyOn instead of fn?

With `spyOn`, you don't need to manually keep the original implementations at somewhere such as
```
      fs.existsSync = originalExistsSync
      fs.unlinkSync = originalUnlinkSync
```

> And the docs also suggests that use jest.spyOn(object, methodName).mockImplementation(() => customImplementation) or object[methodName] = jest.fn(() => customImplementation) to skip the real invocation.

What the docs don't mention is the return of `spyOn` is actually the return of `jest.fn()`, which you can use for both assertions of expected calls and mock return values.

Reply:
It's indeed a sugar of `fn` but the semantic is to call the real method which is emphasized in the docs, I think we should follow the convention.

BTW, I'd like to know the benefit of using `spyOn` instead of `fn` in this case. From my perspective, the only difference is the clean after test case where `spyOn` uses `restore` and `fn` uses assignment.

Reply:
I've checked some articles, it's ok to use `spyOn` here.

Reply:
What I see in your test, it is intended to mock values for the `fs` functions instead of calling through the real `fs` functions, correct? 

AFAIK, if we don't use `spyOn` here, then it has to do the setup you did, which could be quite cumbersome when it comes to maintenance IMHO because managing the original functions at somewhere like this could be a coding overhead for other developers. With these mocks all managed by `jest`, it is much easier to comprehend as we all have the common practice in terms of how to do the mocks.
```
fs.existsSync = originalExistsSync
fs.unlinkSync = originalUnlinkSync
```